subsys/bluetooth/host/id.c

```diff
@@ -701,6 +701,74 @@ void bt_id_pending_keys_update(void)
 	}
 }
 
+static int hci_id_del(const bt_addr_le_t *addr)
+{
+	struct bt_hci_cp_le_rem_dev_from_rl *cp;
+	struct net_buf *buf;
+
+	BT_DBG("addr %s", bt_addr_le_str(addr));
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_REM_DEV_FROM_RL, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	bt_addr_le_copy(&cp->peer_id_addr, addr);
+
+	return bt_hci_cmd_send_sync(BT_HCI_OP_LE_REM_DEV_FROM_RL, buf, NULL);
+}
+
+static int hci_id_overwrite(struct bt_keys *keys)
+{
+	int err = hci_id_add(keys->id, &keys->addr, keys->irk.val);
+
+	switch (err) {
+		case 0: {
+			/* Success */
+			return 0;
+		}
+
+		case -EIO: {
+			/* Delete the identity in the controller's resolve list since it already
+			 * exists to make sure that a potentially new IRK is stored.
+			 */
+			err = hci_id_del(&keys->addr);
```

Comment:
Hmm, -EIO is used to aggregate any controller error. 

The `hci_id_add` operation can in theory return an error for exceeded memory, invalid parameters (this is the one we want to catch I suppose) or if attempting to add a device while advertising, scanning or when a connection is pending. 

So not sure if we can just do this, as only 1 out of 6 error scenarios should do this.

Reply:
Yes, but the BT_HCI_ERR_INVALID_PARAM error (which is returned when a duplicate is found) is being replaced by -EIO in bt_hci_cmd_send_sync. This is why I make hci_id_overwrite try to overwrite the identity using del+add when -EIO is returned.

Reply:
Any ideas?

Reply:
We can add more checks (if not already there) to ensure that the host and controller are in-sync and that the host will not call `hci_id_add` without knowing that it will succeed, thus remove the need for this `case`.

Reply:
I could update bt_hci_cmd_send_sync so it will pass up the BT_HCI_ERR_INVALID_PARAM error instead of replacing it with -EIO? The bluetooth spec says this error code is used for duplicate entries when adding.

The spec does not seem to say anything about what to do if a duplicate is found and it also gives the controller multiple options for when an entry is a duplicate. So the host code has to assume that BT_HCI_ERR_INVALID_PARAM means a duplicate was found, but that the IRK is not checked (Spec allows this).

So the IRK in the resolve list should be overwritten somewhere. This solution does that in the host code. Not how to otherwise avoid an incorrect IRK in the resolve list, since it is loaded with entries from the settings system and therefore contains the previous bonding's IRK.

Reply:
We discussed different approaches to how we can first detect if the IRK has been added on Discord, and I am now updating the PR with a suggestion for how to do this.